Core/Movement: Fix some undermap issues with random movement/fear/blink

Golrag opened the pull request:
<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

- Only move to point if there is a path that is not a shortcut (which will make the unit move through terrain)
- ~Added new function to check if there is a vmap floor without search distance~
- Units that can fly, are underground but far above the vmap floor will stay underground (bronze drakes in tanaris)
- Don't remove PATHFIND_SHORTCUT from path type in some cases


**Target branch(es):**

- 3.3.5

**Issues addressed:** Updates #21255


**Tests performed:**
- Builds
- Blinking around in WSG
- Fearing in WSG
- Fearing in Icecrown (mountains in the centre)
- Blinking in Icecrown

**Known issues and TODO list:**

- ~Unsure about performance. I needed to check if there was a vmap way lower than the default search distance would let me for the drakes in tanaris/cot~


<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Golrag committed:
Core/Movement:
- Only move to point if there is a path that is not a shortcut (which will make the unit move through terrain)
- Added new function to check if there is a vmap floor without search distance
- Units that can fly, are underground but far above the vmap floor will stay underground (bronze drakes in tanaris)
- Don't remove PATHFIND_SHORTCUT from path type in some cases

yvomsx commented:
Vouch +1

Shauren commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -1428,9 +1428,15 @@ void WorldObject::UpdateAllowedPositionZ(float x, float y, float &z) const
+            float const gridHeight = GetMap()->GetGridHeight(x, y) + unit->GetHoverOffset();
+            float const vmapHeight = GetMap()->GetVmapDataFloorZ(x, y, z); // don't add hover offset here. if check might return unintended value
+            // Check if vmap under us
+            if (vmapHeight <= INVALID_HEIGHT)
```
Why did you split these calls? GetMapHeight uses GetMap()->GetHeight which is supposed to return the correct height value, either from grid or vmaps

Golrag replied:
The changes made in Map::GetHeight will make it so that it will select the closest one, not caring if z is lower than grid. The drakes in the caverns of time will rise up again if I use that function.

I'm thinking about reverting that part so we can still use the same function but not sure yet if it could break things again

Shauren commented on src/server/game/Maps/Map.cpp:
```diff
@@ -2441,24 +2441,11 @@ float Map::GetWaterOrGroundLevel(uint32 phasemask, float x, float y, float z, fl
-            vmapHeight = vmgr->getHeight(GetId(), x, y, z, maxSearchDist);
-    }
+    float const mapHeight = GetGridHeight(x, y);
+    float const vmapHeight = checkVMap ? GetVmapDataFloorZ(x, y, z) : VMAP_INVALID_HEIGHT_VALUE;
```
Why replace vmgr->getHeight with vmgr->getAreaAndLiquidData?

Golrag replied:
Was looking for a function that could return the floor without a max search distance, as the default one was not enough so it would return an invalid vmap height iirc

Need to check it again to be sure

Aokromes replied:
@Golrag jump into irc :)

yvomsx replied:
@Aokromes Any news on this? 👍

Aokromes replied:
@yvomsx test it yourself, and give feedback.

yvomsx replied:
@Aokromes Give me a little, will review and post results.

yvomsx commented:
@Golrag Managed to resolve it? i can also take a look for u if u wish.

Killyana commented:
There's an issue about random movements https://github.com/TrinityCore/TrinityCore/issues/21872
You can notice it IG with this npcs: .go c id 23929

yvomsx commented:
Checked this pull, compiled and tested, appears to be working so far.

Persephone commented:
@Golrag Can you work your magic on shadowstep as well? If you try to shadowstep a target that is on a higher ground than yourself, you get placed underground. 

As an example i spawned a dummy (id 32666) in guru here:

http://prntscr.com/m837ek

and when i use shadowstep i end up here:

http://prntscr.com/m837rt

Lorenc95 reviewed (approved):
Tested, working been running 2 days, without a crash or whatsoever, issue is not completely gone i guess but its a major improvement of what it used to be.

IllInuz commented:
@Killyana Above comment is what i meant::

> I found an issue (with blink)

.go xyz -8173.074219 -4719.009766 27.010540 1 4.943924
end position by path finder: -8168.48438 -4738.47656 29.2423859
--> Grid z: invalid
--> vmap z: invalid (we are under vmap)

--> undermap

illfated commented:
>  [ ... ] issue is not completely gone [ ... ]

Yes, I pretty much took it to mean that there are still some bugged coordinates, but it is (of course) better to include actual references to which bugs are still present.

Illuminati2739 commented:
Hello, is this going to be applied or do we have access to to this patch?

Golrag commented:
@Shauren I might have an approach that removes the call to getAreaAndLiquidData. Could just ignore UpdateAllowedPositionZ for flying units, not units who can fly but really are flying. For units that can fly we can keep the code that's currently present in !unit->CanFly() and modify the max_z check.

This would mean that flying units can fly through mountains. While previously they could go all the way to the top and then go back down to reach you.

```diff
--- a/src/server/game/Entities/Object/Object.cpp
+++ b/src/server/game/Entities/Object/Object.cpp
@@ -1422,40 +1422,29 @@ void WorldObject::UpdateAllowedPositionZ(float x, float y, float &z) const
 
     if (Unit const* unit = ToUnit())
     {
-        if (!unit->CanFly())
-        {
-            bool canSwim = unit->CanSwim();
-            float ground_z = z;
-            float max_z;
-            if (canSwim)
-                max_z = GetMapWaterOrGroundLevel(x, y, z, &ground_z);
-            else
-                max_z = ground_z = GetMapHeight(x, y, z);
+        // Unit is flying, ignore.
+        if (unit->IsFlying())
+            return;
 
-            if (max_z > INVALID_HEIGHT)
-            {
-                // hovering units cannot go below their hover height
-                float hoverOffset = unit->GetHoverOffset();
-                max_z += hoverOffset;
-                ground_z += hoverOffset;
-
-                if (z > max_z)
-                    z = max_z;
-                else if (z < ground_z)
-                    z = ground_z;
-            }
-        }
+        bool canSwim = unit->CanSwim();
+        float ground_z = z;
+        float max_z;
+        if (canSwim)
+            max_z = GetMapWaterOrGroundLevel(x, y, z, &ground_z);
         else
+            max_z = ground_z = GetMapHeight(x, y, z);
+
+        if (max_z > INVALID_HEIGHT)
         {
-            float const gridHeight = GetMap()->GetGridHeight(x, y) + unit->GetHoverOffset();
-            float const vmapHeight = GetMap()->GetVmapDataFloorZ(x, y, z); // don't add hover offset here. if check might return unintended value
-            // Check if vmap under us
-            if (vmapHeight <= INVALID_HEIGHT)
-            {
-                // no vmap found, check if unit is under grid height
-                if (z < gridHeight)
-                    z = gridHeight;
-            }
+            // hovering units cannot go below their hover height
+            float hoverOffset = unit->GetHoverOffset();
+            max_z += hoverOffset;
+            ground_z += hoverOffset;
+
+            if (z > max_z && !unit->CanFly())
+                z = max_z;
+            else if (z < ground_z)
+                z = ground_z;
         }
     }
     else
```
note: changes in map::getheight for vmapheight are reverted. still keeping the gridheight change

jackpoz commented:
@Golrag please post the link to a branch, not diff files

Golrag committed:
Core/Object: Ignore UpdateAllowedPositionZ for flying units.

- This will make flying units go through mountains instead of going to the top and back to the bottom to reach you.

Aokromes commented:
@Golrag if this pr is intended to fix also shadow step, still fails

Golrag commented:
I don't know yet what's happening with shadowstep, will try to take a look at it tomorrow. I noticed it goes through MovePositionToFirstCollision without checking if there is a path, but I'm not sure if it returns a good position or not

natnat201193 commented:
if this is of any help, i think it started happening more often when https://github.com/TrinityCore/TrinityCore/pull/21322 was reverted. but i could be wrong.

natnat201193 commented:
Reverted here: https://github.com/TrinityCore/TrinityCore/commit/2dadbda24ac1c2e6ff3dad650c7e25f4b3322342#diff-d64b193f1697462f5500332fb47092ee

Golrag commented:
Could I get some examples of where this issue occurs ? I tried getting it in gurubashi with the example above but could not reproduce it :/ a .gps of the npc would be great

Ovahlord commented:
Well, I've been testing the changes so far and normal terrain just seems to be working fine. wmo terrain however seems to be struggling a bit. The best place to see that is going into Naxxramas and doing a blink towards one of the stairways up. When doing so you should be partially standing inside the stairs. If there wouldn't be a 2nd collision model below the stairs you would still fall straight through them. So yeah. Blink is more or less halfway fixed but still struggles.

Golrag commented:
I'm sorry I just can't reproduce it. I might be missing something..
https://puu.sh/CQWse/07f7ee930a.mp4
https://puu.sh/CQWt6/e97cc7ed50.mp4

Can't know for sure what's going on, but I think that those issues are not in scope of this PR :/
will try again after I updated tools with latest changes

VengefulWoW commented:
The best place is Gurubashi Arena right outside. You can blink into abyss in many different spots. It is bugged as hell. People constantly fall through a map

Illuminati2739 commented:
Is this going to be applied? This fixes blink and shadow step however i do not see this as an actual commit

illfated commented:
The whole point of a Pull Request is to deal with unanswered questions or unsolved issues in the content of the PR before it gets merged. I recommend showing some patience with regards to when this will be merged. It will be merged when reviews and requested changes have been sorted out and solved (and the owners of TrinityCore are happy with the results).

Illuminati2739 commented:
Of course, no rush was intended or referenced in my post, It was a simple question, which was not posed by me or anyone else before hand. Reason for asking was because "All Checks" were completed. 
Thank you

illfated commented:
Very well, fair enough. "All checks have passed" is mostly an indicator that the code passes automatic review and that no obvious errors need to be fixed. TC members are still free to discuss the content and manual testing is needed to see if the in-game behaviour actually stays blizzlike. Real coders will also be able to spot other technical reasons to request changes.

jackpoz commented:
@Shauren do the new changes look better ? can it be merged now ?

Shauren commented on src/server/game/Maps/Map.cpp:
```diff
@@ -2444,14 +2444,7 @@ float Map::GetWaterOrGroundLevel(uint32 phasemask, float x, float y, float z, fl
-            mapHeight = gridHeight;
-    }
-
+    float const mapHeight = GetGridHeight(x, y);
```
Whats the reasoning behind this change? The point of `z` argument is to never return a height that is higher than supplied `z` (and vmgr->getHeight follows this restriction)

Golrag replied:
I noticed some issues with blink without it... might have been because of https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Spells/Spell.cpp#L1368 allowing shortcut paths, not sure if that's supposed to happen. Can try to remove it and change the check to only allow normal paths

jackpoz commented on src/server/game/Movement/PathGenerator.cpp:
```diff
@@ -536,14 +536,14 @@ void PathGenerator::BuildPointPath(const float *startPoint, const float *endPoin
         TC_LOG_DEBUG("maps", "++ PathGenerator::BuildPointPath FAILED! path sized %d returned\n", pointCount);
         BuildShortcut();
-        _type = PATHFIND_NOPATH;
+        _type = PathType(_type | PATHFIND_NOPATH);
```
what does this change btw ?

Golrag replied:
Prevents removing PATHFIND_SHORTCUT from the path type. I agree it's a bit weird to combine it with NOPATH. As most of the checks are already done on NOPATH

Ovahlord commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -1422,33 +1422,28 @@ void WorldObject::UpdateAllowedPositionZ(float x, float y, float &z) const
-            else
-                max_z = ground_z = GetMapHeight(x, y, z);
+        // Unit is flying, ignore.
+        if (unit->IsFlying())
```
You need to revisit the logic in this part here. The changes break hover ground z coordinates for units that can fly and have ground hover movement (for example the valkyrs in ICC). So with the changes made here the valkyrs for example never get a proper ground Z returned because they can fly and they can hover above the ground. The IsFlying check is effectively disabling all hover checks in this case so the valkyr can fly into the ground ignoring the hover offset entirely.

robinsch commented:
WMOs are still bad. For example in Stormwind: https://gyazo.com/a7d9834cb689ffabc5a877c3b169d407

```
.go xyz -8843 634 94
```

Golrag committed:
Core/Object: Revert some changes and let MovePositionToFirstCollision deal with a position without ground

Golrag committed:
Missing groundZ change for objects on transport

Ovahlord commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -3283,10 +3297,27 @@ void WorldObject::MovePositionToFirstCollision(Position &pos, float dist, float
+        if (Unit const* unit = ToUnit())
+        {
+            // flying, ignore.
+            if (unit->IsFlying())
```
better go with  if (unit->CanFly()) because, once again, IsFlying also returns true for units with disabled gravity which are creatures which should still consider ground heights when getting too close to it. (Naming the icc Valykr scenario again)

Golrag committed:
use CanFly instead of IsFlying

Cobus420 commented:
No issue longer with blink and Fear. It works nice here!

illfated commented:
@Cobus420 : Some more details would be helpful, like if you mean this PR content and which commit rev. hash from the TC source you based it on.

jackpoz commented:
is this PR ready to be merged ? could the ones who have been testing it test latest version too and provide some feedback so we know if we can merge it ? thanks 👍

Aokromes commented:
imho yes, it works far better than old code, i don't have got more reports of people falling under floor.

jackpoz commented:
Thanks for the PR :)
